subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaBasePlugin.java

```diff
@@ -143,7 +143,7 @@ private DefaultJavaPluginExtension addExtensions(final Project project) {
         DefaultToolchainSpec toolchainSpec = objectFactory.newInstance(DefaultToolchainSpec.class);
         SourceSetContainer sourceSets = (SourceSetContainer) project.getExtensions().getByName("sourceSets");
         DefaultJavaPluginExtension javaPluginExtension = (DefaultJavaPluginExtension) project.getExtensions().create(JavaPluginExtension.class, "java", DefaultJavaPluginExtension.class, project, sourceSets, toolchainSpec, jvmPluginServices);
-        project.getConvention().getPlugins().put("java", objectFactory.newInstance(DefaultJavaPluginConvention.class, project, javaPluginExtension));
+        DeprecationLogger.whileDisabled(() -> project.getConvention().getPlugins().put("java", objectFactory.newInstance(DefaultJavaPluginConvention.class, project, javaPluginExtension)));
```

Comment:
What is the plan going forward for deprecating the actual convention objects which we are registering here? This PR successfully deprecates `Project#getConvention`, however the `DefaultJavaPluginConvention`, and I am sure many other conventions objects, are still heavily in use -- particularly in dynamic buildscripts. 

I know there is some logic _somewhere_ (I forget where) that allows the field on the java plugin convention to be called on the top-level of the project. For example, see any of these existing tests where `source/targetCompatibility` are called at the root project-level and not on the java extension itself:
https://github.com/gradle/gradle/blob/master/subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/compile/JavaCompileJavaVersionIntegrationTest.groovy#L46-L47
https://github.com/gradle/gradle/blob/master/subprojects/ide/src/integTest/groovy/org/gradle/plugins/ide/eclipse/EclipseWtpWebProjectIntegrationTest.groovy#L35

These are two examples, but there are TONS of examples of this. We are using these conventions everywhere in our tests, particularly older ones, and I'm worried that this PR does not do enough -- or at the very least gets us to the point of "yay we deprecated project.getConvention", while ignoring the fact that our tests (and users) are dynamically using these conventions everywhere in their groovy scripts. 

I have not looked through this PR very much, but I worry that this is just the tip of the iceberg. We should be adding deprecation loggers to all of the conventions objects themselves so that if users are still relying on the dynamic nature of the conventions that they get warnings. Or alternatively, we could add a logger in the path of the dynamic groovy code which allows us to call these conventions methods at the top-level project.